miniapps/adjoint/advection_diffusion.cpp

```diff
@@ -0,0 +1,488 @@
+   //    'ref_levels' of uniform refinement, where 'ref_levels' is a
+   //    command-line parameter. If the mesh is of NURBS type, we convert it to
+   //    a (piecewise-polynomial) high-order mesh.
+   for (int lev = 0; lev < ser_ref_levels; lev++)
+   {
+      mesh->UniformRefinement();
+   }
+
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int lev = 0; lev < par_ref_levels; lev++)
+   {
+      pmesh->UniformRefinement();
+   }
+
+   // 5. Finite Element Spaces
+
+   H1_FECollection fec(1, pmesh->SpaceDimension());
+   ParFiniteElementSpace *fes = new ParFiniteElementSpace(pmesh, &fec);
+
+   HYPRE_Int global_vSize = fes->GlobalTrueVSize();
+   if (myid == 0)
+   {
+      cout << "Number of unknowns: " << global_vSize << endl;
+   }
+
+   // 6. Set up material properties, and primal and adjoint variables
+
+   // p are the fixed material properties
+   Vector p(2);
+   p[0] = 1.0;
+   p[1] = 0.5;
+
+   // U is the size of the solution/primal vector
+   // Set U with the initial conditions
+   ParGridFunction u(fes);
+   FunctionCoefficient u0(u_init);
+   u.ProjectCoefficient(u0);
+
+   cout << "Init u: " << endl;
+   u.Print();
+
+   // TimeDependentOperators need to be TrueDOF Size
+   HypreParVector *U = u.GetTrueDofs();
+
+   // Get boundary conditions
+
+   Array<int> ess_tdof_list;
+   Array<int> essential_attr(pmesh->bdr_attributes.Size());
+   essential_attr[0] = 1;
+   essential_attr[1] = 1;
+   fes->GetEssentialTrueDofs(essential_attr, ess_tdof_list);
+
+   // 7. Setup the TimeDependentAdjointOperator and the CVODESSolver
+   
+   AdvDiffSUNDIALS adv(U->Size(), U->Size(), p, fes, ess_tdof_list);
+
+   // Set the initial time to the TimeDependentAdjointOperator
+   double t = 0.0;
+   adv.SetTime(t);
+
+   // 8. Create the CVODES solver corresponding to the selected step method
+   CVODESSolver *cvodes = new CVODESSolver(fes->GetComm(),
+					   step_mode ? CV_ADAMS : CV_BDF);
+   cvodes->Init(adv);
+   cvodes->UseSundialsLinearSolver();
+
+   // Relative and absolute tolerances for CVODES
+   double reltol = 1e-8, abstol = 1e-6;
+   cvodes->SetSStolerances(reltol, abstol);
+
+   // Initialize adjoint problem settings
+   int checkpoint_steps = 50; ///< steps between checkpoints
+   cvodes->InitAdjointSolve(checkpoint_steps, CV_HERMITE);
+
+   // 9. Perform time-integration for the problem (looping over the time
+   //    iterations, ti,  with a time-step dt).
+   bool done = false;
+   for (int ti = 0; !done; )
+   {
+      double dt_real = max(dt, t_final - t);
+      cvodes->Step(*U, t, dt_real);
+      ti++;
+
+      done = (t >= t_final - 1e-8*dt);
+
+      if (done )
+      {
+         cvodes->PrintInfo(); 
+      }
+   }
+
+   u = *U;
+   cout << "Final Solution: " << t << endl;
+   u.Print();
+
+   // 10. Calculate the quadrature int_x u dx at t = 5
+   // Since it's only a spatial quadrature we evaluate it at t=5
+
```

Comment:
Some comments start with a blank line before the code and some directly above code.